Implement SophiaH & AdaHessian

i404788 opened the pull request:
## Problem (Why?)

Solves https://github.com/kozistr/pytorch_optimizer/issues/173
https://github.com/kozistr/pytorch_optimizer/pull/175 only has SophiaG (which is more difficult to use due to Gauss-Newton-Bartlett estimator)

## Solution (What/How?)

* [x] Implement SophiaH
* [x] Implement AdaHessian
* [x] Refactor AdaHessian to use utility rather than own
* [x] Validate on toy optimization problem

## Other changes (bug fixes, small refactors)

Added utility to compute Hutchinson's approximate Hessian

kozistr commented on pytorch_optimizer/optimizer/adahessian.py:
```diff
@@ -0,0 +1,164 @@
+# Modified from https://github.com/davda54/ada-hessian/blob/master/ada_hessian.py (MIT David Samuel)
+
+
+class AdaHessian(Optimizer, BaseOptimizer):
```
Actually, there was `AdaHessian optimizer` before #59, but I removed it because there were issues with calculating Rademacher distribution and autograd stuff on the recent pytorch (maybe >=1.9).

I guess we need an alternative version that works on the recent pytorch.

kozistr replied:
cool! im looking forward to your work!

kozistr reviewed (commented):
wow! thanks for the contribution :) 

It's draft PR yet, but I just leave a comment about AdaHessian!

+) Don't worry about CI, I can handle it later :)

i404788 commented:
Early results are promising:
![image](https://github.com/kozistr/pytorch_optimizer/assets/50617709/116294be-d114-4dad-9370-8a305a53cf8a)

It seems like the `torch.autograd.grad` version works just fine with Gaussians, I'll see if I can get Rademacher to work as well as another option to tune

mailgpa commented:
Installed package from git: https://github.com/i404788/pytorch_optimizer

Got an error with nanoGPT, tried to replace pytorch AdamW with SophiaH

```
    def configure_optimizers(self, weight_decay, learning_rate, betas, device_type):
        # start with all of the candidate parameters
        param_dict = {pn: p for pn, p in self.named_parameters()}
        # filter out those that do not require grad
        param_dict = {pn: p for pn, p in param_dict.items() if p.requires_grad}
        # create optim groups. Any parameters that is 2D will be weight decayed, otherwise no.
        # i.e. all weight tensors in matmuls + embeddings decay, all biases and layernorms don't.
        decay_params = [p for n, p in param_dict.items() if p.dim() >= 2]
        nodecay_params = [p for n, p in param_dict.items() if p.dim() < 2]
        optim_groups = [
            {'params': decay_params, 'weight_decay': weight_decay},
            {'params': nodecay_params, 'weight_decay': 0.0}
        ]
        num_decay_params = sum(p.numel() for p in decay_params)
        num_nodecay_params = sum(p.numel() for p in nodecay_params)
        print(f"num decayed parameter tensors: {len(decay_params)}, with {num_decay_params:,} parameters")
        print(f"num non-decayed parameter tensors: {len(nodecay_params)}, with {num_nodecay_params:,} parameters")
        # Create AdamW optimizer and use the fused version if it is available
        # fused_available = 'fused' in inspect.signature(torch.optim.AdamW).parameters
        # use_fused = False # fused_available and device_type == 'cuda'
        # extra_args = dict(fused=True) if use_fused else dict()
        # optimizer = torch.optim.AdamW(optim_groups, lr=learning_rate, betas=betas, **extra_args)
        optimizer = SophiaH(optim_groups, lr=learning_rate, betas=betas)
        # print(f"using fused AdamW: {use_fused}")

        return optimizer
```

```
Traceback (most recent call last):
  File "/home/gpa/dev/nanoGPT/train.py", line 307, in <module>
    scaler.step(optimizer)
  File "/home/gpa/.local/lib/python3.11/site-packages/torch/cuda/amp/grad_scaler.py", line 315, in step
    return optimizer.step(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gpa/.local/lib/python3.11/site-packages/torch/optim/optimizer.py", line 280, in wrapper
    out = func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gpa/.local/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gpa/.local/lib/python3.11/site-packages/pytorch_optimizer/optimizer/sophiah.py", line 76, in step
    self.compute_hutchinson_hessian(self.n_samples)
  File "/home/gpa/.local/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gpa/.local/lib/python3.11/site-packages/pytorch_optimizer/base/optimizer.py", line 47, in compute_hutchinson_hessian
    h_zs = torch.autograd.grad(grads, params, grad_outputs=zs, retain_graph=i < nsamples - 1)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gpa/.local/lib/python3.11/site-packages/torch/autograd/__init__.py", line 319, in grad
    result = Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
             
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```

mailgpa commented:
Thanks. It now fails with "aten::_scaled_dot_product_flash_attention_backward is not implemented", but it's not related to the optimizer, apparently

i404788 commented:
There have been a few changes since last time:
- SignSGD added (since SophiaH is SignSGD when hessian is negative I wanted to compared)
- Modified tests to allow for hessian optimizers (Please check if the changes make sense)
- Added an option in `.step()` to allow for providing your own hessian source (this way you can make SophiaG without any modifications)
- Modified the hyperparameters to match the paper
- Allow selecting the distribution used in Hutchinson Estimator (`gaussian` & `rademacher` supported)
- General bug fixes

I've tested all of the optimizers on an example which is supposedly the best-case for SophiaH:
![image](https://github.com/kozistr/pytorch_optimizer/assets/50617709/f7f180e6-b143-4815-bc53-b1a2ed558d35)

```
Adam    TaskResult(best_loss=0.0109736305, last_loss=0.07625367)
SGD     TaskResult(best_loss=0.0049692118, last_loss=0.0049692118)
SophiaH TaskResult(best_loss=7.2648654e-06, last_loss=7.2648654e-06)
SignSGD TaskResult(best_loss=0.009993456, last_loss=0.019895213)
```

The behavior of SophiaH is similar to the behavior described in the paper, but Adam/SGD aren't as bad as described in the paper, this might be due to the example being quite large in magnitude, or because my example is not quite the same.

kozistr reviewed (commented):
great work!

one question is, does AdaHessian with Rademacher distribution work with Pytorch 1.x and 2.x? It seems that the implementation is almost identical to the original and my previous implementations.

and don't worry about lint and test coverage. I'll handle it after merging this PR!

i404788 commented:
Yes it does seem to work for `1.13.1` & `2.0.1`. 

Some details on the stability/accuracy of the hessian calculation:

*There is a difference between the distributions (at least for adahessian) but both work*
![image](https://github.com/kozistr/pytorch_optimizer/assets/50617709/91c78ee3-8b05-485e-8fb4-bedfd52b31e6)

*Adahessian does have sometimes instability with gaussian, I think it's because Adahessian relies on the hessian too much without any clipping*
![image](https://github.com/kozistr/pytorch_optimizer/assets/50617709/35e53998-a1cd-4c83-bbb3-bb6bfcf3fc51)

*SophiaH does not have any instability in my testing:*
![image](https://github.com/kozistr/pytorch_optimizer/assets/50617709/3440e703-7383-4ccd-be33-3ca04b3010ca)

* Using HVP to calculate the hessian instead of grad doesn't make much of a difference:
![image](https://github.com/kozistr/pytorch_optimizer/assets/50617709/b33a4913-0f60-450d-9abd-d986867fd533)

kozistr commented:
> Yes it does seem to work for `1.13.1` & `2.0.1`.
> 
> Some details on the stability/accuracy of the hessian calculation:
> 
> _There is a difference between the distributions (at least for adahessian) but both work_ ![image](https://user-images.githubusercontent.com/50617709/243996559-91c78ee3-8b05-485e-8fb4-bedfd52b31e6.png)
> 
> _Adahessian does have sometimes instability with gaussian, I think it's because Adahessian relies on the hessian too much without any clipping_ ![image](https://user-images.githubusercontent.com/50617709/243997370-35e53998-a1cd-4c83-bbb3-bb6bfcf3fc51.png)
> 
> _SophiaH does not have any instability in my testing:_ ![image](https://user-images.githubusercontent.com/50617709/243999251-3440e703-7383-4ccd-be33-3ca04b3010ca.png)
> 
> _Using HVP to calculate the hessian instead of grad doesn't make much of a difference:_ ![image](https://user-images.githubusercontent.com/50617709/244003826-b33a4913-0f60-450d-9abd-d986867fd533.png)
> 
> HVP code:
> 
> ```
> def hessian_estimation(model, task: Task):
>     from jax.tree_util import tree_map
>     # from functorch import vmap
> 
>     func, params = make_functional(model)
> 
>     def run_model(params):
>         # NOTE: normally you would inject data to use in MSE/BCE/etc through closure
>         return task.loss_f(func(params))
> 
>     noise = tree_map(lambda p: torch.randint(0, 2, p.size(), dtype=p.dtype, device=p.device) * 2.0 - 1.0, params)
>     # noise = tree_map(lambda v: torch.randn_like(v), params)
>     # loss_, hvp_est = tree_map(lambda v: v.t(), vhp(run_model, params, noise))
>     # loss_, hvp_est = hvp(run_model, params, noise)
>     loss_, hvp_est = jvp(grad(run_model), (params,), (noise,))
>     return tree_map(lambda a, b: a*b, hvp_est, noise)
> 
> ...
> 
> if use_hvp_hessian:
>     if opt._step % opt.update_period == 0:
>         hessian = hessian_estimation(model, task)
>         opt.step(hessian=hessian)
>     else:
>         opt.step()
> else:
>     opt.step()
> ```

cool! thanks for the hard work! I'll merge this PR and do some refactors.
I'll let you know when new version with Sophia optimizer is deployed

thank you!

kozistr commented:
I just deployed [v2.10.0](https://github.com/kozistr/pytorch_optimizer/releases/tag/v2.10.0). thank you!
